### ISSUE TITLE
[QUESTION] CPU name logic

### ISSUE THREAD

**Issue by RedBearAK** [aristocratos/bpytop#68]
Is the following code block the entire logic in bpytop determining the CPU name that gets displayed? 

I'm just curious at this point because I have a CPU that reports as "Intel(R) Core(TM)2 Duo CPU" and bpytop just ends up displaying "Duo", which is not that helpful since it could also refer to a Core Duo CPU. 

I'm wondering if anyone out there has already built a function to more reliably return a useful and accurate (but shortened) CPU name that would better suit the bpytop use case. Failing that, maybe I can add to the filtering logic here. For instance, filtering out "Processor" from "ARMv7 Processor rev 3 (v7l)" on the Raspberry Pi. 

This function did pretty good with a Ryzen 3600X processor, but maybe that was just because it called itself "Ryzen 3600X". I didn't check. 

```
def get_cpu_name() -> str:
	'''Fetch a suitable CPU identifier from the CPU model name string'''
	name: str = ""
	nlist: List = []
	command: str = ""
	cmd_out: str = ""
	rem_line: str = ""
	if SYSTEM == "Linux":
		command = "cat /proc/cpuinfo"
		rem_line = "model name"
	elif SYSTEM == "MacOS":
		command ="sysctl -n machdep.cpu.brand_string"
	elif SYSTEM == "BSD":
		command ="sysctl hw.model"
		rem_line = "hw.model"


	try:
		cmd_out = subprocess.check_output("LANG=C " + command, shell=True, universal_newlines=True)
	except:
		pass
	if rem_line:
		for line in cmd_out.split("\n"):
			if rem_line in line:
				name = re.sub( ".*" + rem_line + ".*:", "", line,1).lstrip()
	else:
		name = cmd_out
	nlist = name.split(" ")
	if "Xeon" in name and "CPU" in name:
		name = nlist[nlist.index("CPU")+1]
	elif "Ryzen" in name:
		name = " ".join(nlist[nlist.index("Ryzen"):nlist.index("Ryzen")+3])
	elif "CPU" in name and not nlist[0] == "CPU":
		name = nlist[nlist.index("CPU")-1]


	return name
```

**Comment by RedBearAK** [comment]
I feel like you could get a long way toward reliably displaying a short, accurate CPU model name by just striking these patterns and maybe a few others from the retrieved name: 

`Processor`
`CPU`
`(R)`
`(TM)`

On my systems that would result in: 

`Intel Core2 Duo` 

and

`ARMv7 rev 3 (v7l)`

Pretty clean already without too much trouble. It might even make sense to strike pretty much anything enclosed in parentheses or brackets, regardless of what it is. But I think just removing the above patterns by default would clean up an awful lot of common CPU model names to the point that what bpytop has room to display would end up looking pretty nice in many cases.

**Comment by aristocratos** [comment]
This function is never gonna be perfect for all possible naming schemes without making it way too long. That's the reason I included the "custom_cpu_name" option for anyone not liking what it automatically picked up.

But sure I can add some more cleanup to the name with the strings you recommended.

But it wouldn't fix "Intel(R) Core(TM)2 Duo CPU" not being reported as just "Duo" since that information has already been lost after:
``` python
elif "CPU" in name and not nlist[0] == "CPU":
		name = nlist[nlist.index("CPU")-1]
```
which gets correct naming for most i3-9 Intel cpus.

Would need a separate if statement for just Intel Core Duo cpus, which to be fair a very small subset of users is likely to have.

**Comment by RedBearAK** [comment]
That's true, it would be a deep rabbit hole to try and cover everything out there. I'm not interested in adding more than a few minor cleanup tweaks to cover the most common cases. 

If you like the code block above for all the Intel i[N] processors, the simplest suggestion I can think of is to simply bypass that block if you don't detect "i[N]" in the model name (if you know what I mean). Then the Duos would just get cleaned up like in my example earlier. You shouldn't even need a separate if statement. Just exclude anything that doesn't have i3/i5/i7/i9. 

Although, technically, if you don't also strip everything after "CPU" my C2D would also display "T7500 @ 2.2GHz" at the end of the model name. 

I wasn't sure about this earlier but I think it would usually be fine to also add "Intel" and other brand names to the list of patterns to remove, since the brand will normally be quite apparent from the rest of the model name. 

In that case my C2D would show as: 

`Core2 Duo T7500 @ 2.2GHz`

and if you strip everything after the "CPU" pattern of course it would just be: 

`Core2 Duo`

If I ever get familiar enough with python to attempt my own patch for this I will try instead to strip out the GHz numbers and "@" symbol and just leave: 

`Core2 Duo T7500`

But that would probably be too much trouble for you to spend time on. 

Something to think about: 

I think you may underestimate the number of pre-i-series Intel CPUs out there. With Apple machines alone there is a huge block (millions) of older iMacs, MacBooks and Mac Pros that are still chugging along but will soon be incompatible with even the DosDude1 patched versions of macOS that allowed me to put Mojave on machines like my 2011 MacBook Pro. As it becomes more and more difficult and pointless to try and upgrade those systems to more current versions of macOS, many of those machines are being, and will continue be, revived with Linux. 

Even the two 2007 iMacs and 2007 MacBook in my house support 6GB of RAM, which means they will continue to be very usable computers for years to come. And these old Apple models are almost all 100% 64-bit systems, including a 64-bit UEFI BIOS. So there is very little holding them back from running future versions of Linux even as Linux distros are busy abandoning 32-bit systems. Core Duo systems are kind of a cutoff point in that sense. 

In other words, there won't be fewer of these Duo-based machines in the future. There will be a flood of them entering the Linux user base over the next decade as their last compatible macOS version loses support for software like web browsers, and the owners finally decide sticking with macOS isn't as useful as moving to Linux.

**Comment by RedBearAK** [comment]
You really shouldn't close an issue before we have a chance to heap praise on your resolution of the problem. 

MacBook now says "Core2 Duo       T7500", which is much more than I expected to see. 

Raspberry Pi 4B now says "ARMv7 rev 3 (v7l)" as expected. Very nice. 

10/10

**Comment by RedBearAK** [comment]
It wasn't apparent from the post above because I failed to use the code quotes, but the actual CPU name for my Duo has a ton of spaces between the name and model number. Sort of like this: 

```Core2 Duo . . . . T7500```

(Damn, even with the code quotes it's not rendering the multiple spaces properly when I preview. I'll have to add periods or something just to get the spacing right. So the periods show above don't actually exist in the bpytop output.)

Not a big deal but I kept wondering why you would separate them like that. I figured out that your code is still so simple that it doesn't have a step for filtering out excess whitespace, and there is a ton of excess whitespace in the Duo model name. I made a simple change to my bpytop code and it seems to work, and as far as I understand (at the moment) it shouldn't interfere with the other "replace" stuff just below it, because it should leave exactly one space after each element in the name. 

You can see the line I added here just before the "return" line. I don't know if that is precisely the right place to put it. Please look it over and see if you want to add it. It's this line: 

```
	name = " ".join(name.split())
```

The whole code block: 

```
def get_cpu_name() -> str:
	'''Fetch a suitable CPU identifier from the CPU model name string'''
	name: str = ""
	nlist: List = []
	command: str = ""
	cmd_out: str = ""
	rem_line: str = ""
	if SYSTEM == "Linux":
		command = "cat /proc/cpuinfo"
		rem_line = "model name"
	elif SYSTEM == "MacOS":
		command ="sysctl -n machdep.cpu.brand_string"
	elif SYSTEM == "BSD":
		command ="sysctl hw.model"
		rem_line = "hw.model"

	try:
		cmd_out = subprocess.check_output("LANG=C " + command, shell=True, universal_newlines=True)
	except:
		pass
	if rem_line:
		for line in cmd_out.split("\n"):
			if rem_line in line:
				name = re.sub( ".*" + rem_line + ".*:", "", line,1).lstrip()
	else:
		name = cmd_out
	nlist = name.split(" ")
	if "Xeon" in name and "CPU" in name:
		name = nlist[nlist.index("CPU")+1]
	elif "Ryzen" in name:
		name = " ".join(nlist[nlist.index("Ryzen"):nlist.index("Ryzen")+3])
	elif "Duo" in name and "@" in name:
		name = " ".join(nlist[:nlist.index("@")])
	elif "CPU" in name and not nlist[0] == "CPU":
		name = nlist[nlist.index("CPU")-1]

	name = " ".join(name.split())
	
	return name.replace("Processor ", "").replace("CPU ", "").replace("(R)", "").replace("(TM)", "").replace("Intel ", "")
```

**Comment by RedBearAK** [comment]
Did my best to make a proper pull request with the suggested change.

**Comment by aristocratos** [comment]
Merged in v1.0.21

**Comment by viv-4** [comment]
Ran into a related issue preventing bpytop from starting:
```
line 4216, in get_cpu_name
    name = nlist[nlist.index("CPU")+1]
IndexError: list index out of range
```

`cat /proc/cpuinfo | grep "model name"`
```
model name      : Intel(R) Xeon(R) CPU
model name      : Intel(R) Xeon(R) CPU
model name      : Intel(R) Xeon(R) CPU
model name      : Intel(R) Xeon(R) CPU
```

Changed it to
`name = nlist[nlist.index("CPU")-1]`
which fixed the issue and reports the cpu name as `Xeon`

Not sure what input it was expecting so not sure what the best fix here is.
Perhaps a check for "CPU" at the end of the string but that is just adding more `if`s
```
        if "Xeon" in name and "CPU" in name:
                if nlist.index("CPU") == len(nlist)-1:
                        name = nlist[nlist.index("CPU")-1]
                else:
                        name = nlist[nlist.index("CPU")+1]
```

**Comment by aristocratos** [comment]
@viv-4 
Fixed in v1.0.32
